src/qmapshack/gis/rte/router/CRouterOptimization.cpp

```diff
@@ -0,0 +1,331 @@
+            {
+                bestBeginIndex = beginIndex;
+                bestEndIndex = endIndex;
+                bestTwoOptGain = newRangeCosts-oldRangeCosts;
+            }
+        }
+    }
+
+    newOrder = SGisLine(oldOrder);
+    if(bestEndIndex >= 0 && bestBeginIndex >= 0)
+    {
+        std::reverse(newOrder.begin()+bestBeginIndex, newOrder.begin()+bestEndIndex);
+    }
+    return bestTwoOptGain;
+}
+
+qreal CRouterOptimization::getRealRouteCosts(const SGisLine &line, qreal costCutoff)
+{
+    qreal costs = 0;
+    for(int i = 0; i < line.length()-1; i++)
+    {
+        const routing_cache_item_t* route = getRoute(line[i].coord, line[i+1].coord);
+        if(route == nullptr)
+        {
+            return -1;
+        }
+        costs += route->costs;
+
+        if(costCutoff > 0 && costs > costCutoff)
+        {
+            return -1;
+        }
+    }
+    return costs;
+}
+
+qreal CRouterOptimization::bestKnownDistance(const IGisLine::point_t& start, const IGisLine::point_t& end)
+{
+    //10 digits after the decimal point in exponential format should be by far enough
+    QString start_key=QString::number(start.coord.x(), 'e', 10) + QString::number(start.coord.y(), 'e', 10);
+    QString end_key=QString::number(end.coord.x(), 'e', 10) + QString::number(end.coord.y(), 'e', 10);
+
+    if(!routingCache.contains(start_key))
+    {
+        routingCache[start_key] = QMap<QString, routing_cache_item_t>();
+    }
+
+    if(routingCache[start_key].contains(end_key))
+    {
+        return routingCache[start_key][end_key].costs;
+    }
+    else
+    {
+        //Multiply it with the average of the minimum occuring factor and the average factor
+        // to get a reasonable compromise of optimization speed and optimality of results
+        if(totalNumOfRoutes > 0 && minAirToCostFactor > 0)
+        {
+            return GPS_Math_DistanceQuick(start.coord.x(), start.coord.y(),
+                                          end.coord.x(), end.coord.y())
+                   * (totalAirToCosts/totalNumOfRoutes + minAirToCostFactor)/2;
+        }
+        else
+        {
+            return GPS_Math_DistanceQuick(start.coord.x(), start.coord.y(),
+                                          end.coord.x(), end.coord.y());
+        }
+    }
+}
+
+const CRouterOptimization::routing_cache_item_t* CRouterOptimization::getRoute(const QPointF& start, const QPointF& end)
+{
+    QString start_key=QString::number(start.x(), 'e', 10) + QString::number(start.y(), 'e', 10);
+    QString end_key=QString::number(end.x(), 'e', 10) + QString::number(end.y(), 'e', 10);
+
+    if(!routingCache.contains(start_key))
+    {
+        routingCache[start_key] = QMap<QString, routing_cache_item_t>();
+    }
+
+    if(!routingCache[start_key].contains(end_key))
+    {
+        routing_cache_item_t cacheItem;
+        int response = CRouterSetup::self().calcRoute(start, end, cacheItem.route, &cacheItem.costs);
```

Comment:
Why not pass a reference to cacheItem?

Reply:
I wanted to avoid to duplicate the code in `calcRoute` for the passing `routing_cache_item_t` directly.

Reply:
ok